### ISSUE TITLE
ENH: interface for different geometry engines through GeometryDtype/Array subclasses

### ISSUE THREAD

**Issue by jorisvandenbossche** [geopandas/geopandas#2769]
There have been some discussions recently on the topic of having different "engines" for the actual geometries and geospatial operations (i.e. shapely in current geopandas). Example use cases would be to use geographical objects from [spherely](https://github.com/benbovy/spherely), or to use [cuspatial](https://github.com/rapidsai/cuspatial/) to run the geospatial operations on the GPU.

One potential way to do this would be to allow subclasses of a base geo dtype/array:

* We define an (abstract) base class for the dtype and array that define all the methods we expect / need for operations in geopandas (for example `BaseGeoDtype` / `BaseGeoArray`)
  * (only subclassing the dtype class could actually be sufficient) 
  * (subclasses wouldn't necessarily be expected to implement all methods we have for the shapely backend, they can raise NotImplementedError for non-essential ones, which bubbles up to the user)
* Our own `GeometryDtype` / `GeometryArray` would be a concrete subclass that implements this using `shapely`
* In the GeoSeries / GeoDataFrame constructors, we would allow any such subclass (instead of only GeometryDtype/Array)
  * The fact that it is a subclass is the way we recognize that the passed data is "geo" data and thus fitting for a GeoSeries or geo column in a GeoDataFrame
  * Our GeoSeries / GeoDataFrame methods already dispatch for (almost?) all functionality to a method on the array, so as long as the array object has the expected methods, this would keep working 
* Such subclasses could be implemented both in GeoPandas itself as in external projects

Some aspects of this option:

* Is a separate array/dtype class needed, or could it be done with reusing the existing dtype and array class? Assuming that another project would provide a module with the same namespace as shapely and geometry object with the same methods, you could almost directly use our implementation without having to reimplement it yourself (if there is some way to register / provide this namespace). But I am not sure this would actually be easier, and it also forces you to be exactly compatible with shapely. 
* Is a subclass needed? Strictly speaking not, as long as we have some way to "recognize" a geo dtype. This could also be a certain interface (expected methods being present, i.e. duck typing). But, since the pandas extension dtypes already work with subclassing (you need to subclass pandas' ExtensionDtype), it seems logical to have a geo extension subclass a base geo extension dtype (which could be a slim subclass of pandas' base ExtensionDtype)
* This solution allows to have different geometry columns with different dtypes in a single GeoDataFrame (eg projected geometry and geography)
* Since the spatial index lives on the array object, this solution should also allow external engines to use their own spatial index implementation
* There are probably some places where we might call shapely functionality directly, instead of through a methods on the array. For those cases, we could add more methods to the GeoArray where needed.

**Comment by ShaneZhengNZ** [comment]
I am very keen to see some further discussion around this. Whether GeoPandas will be compatible with GPU accelerated libraries, such as CuDF, etc.

**Comment by martinfleis** [comment]
Hi @ShaneZhengNZ,
That is a good question. It is unlikely that `GeoDataFrame` will be backed by `cuDF` unless this will somehow get implemented in `pandas`. But what was discussed as an option is to have `GeometryArray` depend on `cuSpatial`. But all of that is just a theory now and no plans have been made. It also depends a lot on development of `cuSpatial` itself. We would probably just provide a way of plugging that in.

**Comment by jorisvandenbossche** [comment]
The above description is mostly about the implementation side of things, but probably a more important aspect (and which will also guide the implementation) is how we want this to work for the user.

In any case, we need different array classes, but those could be linked to a single, parametrized `GeometryDtype`, or to each their own geo dtype subclass. So specifying the dtype could either be (using a specialized point dtype as example):

- Parametrized: `dtype=geopandas.GeometryDtype(storage="point")` (or other keyword name, just for illustration), and then potentially with a string alias like `dtype="geometry[point]"`
- Subclass: `dtype=geopandas.PointDtype()`, with a potential string alias like `dtype="point"`

That's for explicitly specifying the dtype. But we will also want some way for the user to specify some default dtype to use (depending on the case, but for example to use spherely by default for lon/lat data, or to use PointDtype by default when the input file/data uses geoarrow.point). We have an option machinery `geopandas.options` (defined in `geopandas/_config.py`) that we could use for this.

**Comment by martinfleis** [comment]
I honestly can't tell which is better, if parametrisation or subclassing. Both are fine I suppose, though subclassing may leave less space to confusion. Right now you can do `.astype('geometry')` and it will do what is intended. With subclassing, this can lead to confusions as you would need to distinguish between `"geometry[point]"` and `"geometry[shapely]"` (or whatever) with the default being unclear. If you, for example, have an array of spherely geoms with the `geometry[spherical]` dtype and it, for some reason ends up with the object dtype, `.astype('geometry')` would need to infer correct type or something. Or raise, which is a point of friction. Having `GeographyDtype` or similar and doing `.astype('geography')` feels less error prone.

**Comment by m-richards** [comment]
Also no strong feelings. These are quick thoughts as I haven't had much time to think about it. I do think that we have two dimensions to accomodate and that makes sense to reflect in the structure (planar geometry vs spherical) and then memory representation (geoarrow vs wrappers to c pointers) and then shapely/ spherely effectively specify something about both dimensions. If you were to do `geometry[point], geometry[shapely], geometry[spherely] etc` there's no good way to represent a hypothetical geoarrow layout calculations on a spheriod. So something like a subclass per planar/spheriod and a parameter to reflect the underlying layout/ engine makes sense to me. Whatever we pick, we want to make sure that a naive astype("geometry") still works. We're also going to want to think about what the GeoDataFrame constructor semantics look like, do we need a new keyword, or do we expect that a user will set up a GeoSeries right / do the geometry type / layout casting after the fact.

I'm not very up to date, but I recall reading there's been back and forth in pandas on the parametrised string dtype backends (object, arrow and maybe now the numpy strings too?) but it's probably worth thinking about if there are any lessons to learn from that.

**Comment by jorisvandenbossche** [comment]
Some late notes that I took during the last community meeting:

Reasons to go for a parametrized `GeometryDtype`:

- It keeps the "geometry" name, which is already present in many of our APIs (`.geometry` attribute, `set_geometry`/`rename_geometry`, keywords with that name, etc). 
- This might be easier to support a global toggle (to enable using the spherical support whenever loading data with geographic coordinates), because at that point you might not always know in advance which type you get when loading data. But when you know it is always a "geometry" dtype, then that might be easier to adapt.

Reasons to not use the parametrized dtype:

- Less explicit or visible. Also, what would we then show in the dtype repr (e.g. when checking `gdf.dtypes`)? Just "geometry", or then always `"geometry[planar]"` or `"geometry[spherical]"` (but then also people that would just always keep using the default geometry would start constantly seeing that)
- Keep the backend or engine keyword to indicate the "compute".

---

I have a POC for spherely support in https://github.com/geopandas/geopandas/pull/3520. While I did use the parametrized dtype approach, I think I might now actually be rather leaning to the subclass. 

One question raised by implementing it this way: which values do we use for the `engine` keyword? In the POC I used the values "planar" (default) and "spherical". But do we want to describe the type of the engine (planar vs spherical) or the package being used (shapely vs spherely)? I like the type more, but at some point we might also have more than one engine for planar geometries. 
This is essentially also the point that @m-richards is making in the comment above: mixing the two dimensions (type of engine and the actual engine implementation) in a single keyword is probably not a good idea. Of course we could also have two keywords in a parametrized GeometryDtype (like `type="spherical", engine="spherely"`), but not sure I like that.

**Comment by jorisvandenbossche** [comment]
Another argument for going with the subclasses (instead of a parametrized GeometryDtype) is that we then could use the "geography" name for this. While there might also be reasons to _not_ do this, in practice almost any other system (or at least databases) that supports spherical/geographic coordinates like PostGIS, BigQuery, Snowflake, DuckDB, etc use "geometry" vs "geography" types.

**Comment by m-richards** [comment]
>One question raised by implementing it this way: which values do we use for the engine keyword? In the POC I used the values "planar" (default) and "spherical". But do we want to describe the type of the engine (planar vs spherical) or the package being used (shapely vs spherely)? I like the type more, but at some point we might also have more than one engine for planar geometries.

I think you've explained this better than I did. I agree that it makes more sense to specify the type of geometry calculation model (I think it's good a user can say they want planar / spherical geometry and we can pick a sane default, rather than them having to be aware that that spherical support comes from spherely). But this does lead you towards having parametrized dtypes with two keywords if you do have multiple compute engines. Maybe this isn't great from a consistency perspective, but perhaps there's a way to register a dummy dtype such that `astype("geometry[planar]")` works in the naive case, but then an advanced user could choose to specify `astype("geoarrow-rs")` as shorthand which implies both the compute engine and the geometry model (at least currently libraries aren't supporting multiple geometry models, but I suppose that's not guaranteed to be true). 

As you mention in the PR, we could also have `to_spherical` as conversion methods. We could also opt to expose the compute engine as a geopandas only concept, that you do via these conversion methods or when data is read, which has the benefit and drawback of not integrating with the pandas dtypes.

> Another argument for going with the subclasses (instead of a parametrized GeometryDtype) is that we then could use the "geography" name for this. While there might also be reasons to not do this

It does work and has precedent, though it's less descriptive than type=planar/spherical. I think it's slightly more awkward in our case due to existing terminology `active_geometry_name, set_geometry, rename_geometry`. It'd be nice if we could treat "geometry" as an umbrella term where planar and spherical are more specific, but maybe that's not realistic.

**Comment by jorisvandenbossche** [comment]
Apologies with being slow to update the issue here, but we discussed this a bit more in the last community (video) meeting, and there the general sentiment was to stick to the terminology "geometry" for geopandas (and then distinguish based on planar vs spherical geometries).

We didn't take that detailed notes, so trying to summarize some of the discussion based on my recollection (but to others, please correct or complement where necessary):

- Although it is used in the database world, the term "geography" can be a mismatch / confusing compared to the _study field_ of geography (concrete example: Martin is a geographer (correct?) but for _his_ work in geographic data science, he would typically be using the geometry data type ..)
- Using "geometry" and "geography" for the dtypes leads to awkward formulation when talking about the "geometry objects" (the points, lines, polygons, etc). Also in a geography/spherical context, it are geometry objects (for example in the Parquet spec about the native geospatial types, the term "geometry" or "geometry object" was replaced with "geospatial instance" after the comment that "geometry" was confusing given the spec introduced both a geometry and geography data type)
- The s2geometry library itself talks about "geometries" (it does use "geographic" in its documentation, but to refer to "geographic datasets" for which s2 can be used)
- In addition to what is mentioned above about the geopandas API all using "geometry", it would also be nice to keep `obj.dtype == "geometry"` working for all possible types that geopandas provides (and that would also make switching to use spherical for lon/lat a smoother transition). It is possible to do that with the subclasses as well, but it is certainly cleaner with a single parametrized GeometryDtype.

**Comment by benbovy** [comment]
Thanks @jorisvandenbossche for the summary.

I think those are all fair points. Naming is hard and various choices have been already made upstream so I'm afraid there's no straightforward and unanimous answer to this problem.

Sorry I wasn't there at the last community meeting but let me take some counter points (I hope it won't be counter productive):

> Although it is used in the database world, the term "geography" can be a mismatch / confusing compared to the study field of geography (concrete example: Martin is a geographer (correct?) but for his work in geographic data science, he would typically be using the geometry data type ..)

Agreed that confusion is unfortunate but in practice it might be better to prioritize consistency within the software / database ecosystem (PostGIS, BigQuery, Snowflake, DuckDB... as you mentioned above) over "pure" terminology.

> the Parquet spec about the native geospatial types, the term "geometry" or "geometry object" was replaced with "geospatial instance" after the comment that "geometry" was confusing given the spec introduced both a geometry and geography data type)

Yes but IIUC the Parquet specs still introduced explicit GEOMETRY and GEOGRAPHY logical types, which would naturally  translate to GeometryDtype and GeographyDtype?

> The s2geometry library itself talks about "geometries" (it does use "geographic" in its documentation, but to refer to "geographic datasets" for which s2 can be used)

Yes but now we also have s2geography and spherely, which talk about "geography" and are closer to GeoPandas in the upstream dependency chain. Especially Spherely that we can use alongside GeoPandas within the same Python interactive session.

> it would also be nice to keep obj.dtype == "geometry" working for all possible types that geopandas provides (and that would also make switching to use spherical for lon/lat a smoother transition)

Is smooth transition a real issue given that geopandas doesn't support spherical objects yet? What about the opt-in, backward compatible suggestion below?

- obj.dtype == "geometry" returns True for planar types (as currently with all shapely-based objects)
- obj.dtype == "geography" returns True for spherical types (once supported)
- obj.dtype == "geospatial" returns True for all possible types that geopandas provides (I picked "geospatial" from the Parquet specs but it could be whatever name chosen for the base type common to geometry and geography)

---

In short, personally I also prefer the terms "planar geometry" vs. "spherical geometry" over "geometry" vs. "geography", but it looks like the ecosystem (databases, Parquet, Spherely, etc.) evolved mostly towards adopting the latter conventions and it might be weird for GeoPandas not following it?

**Comment by martinfleis** [comment]
We also have to keep in mind our existing API. Our `set_geometry`, `active_geometry_name`, `rename_geometry` methods and `geometry` keywords (e.g. in the constructor) need to make sense when spherical support lands. And if we start talking about spherical geometries as geographies, the internal API consistency of geopandas falls apart.

> Yes but now we also have s2geography and spherely, which talk about "geography" and are closer to GeoPandas in the upstream dependency chain. Especially Spherely that we can use alongside GeoPandas within the same Python interactive session.

We also discussed a proposal to make then change in spherely to stick closer to the S2 terminology rather than introducing the term geography but I don't think Joris managed to open an issue about that yet.

**Comment by benbovy** [comment]
> We also have to keep in mind our existing API. Our set_geometry, active_geometry_name, rename_geometry methods and geometry keywords (e.g. in the constructor) need to make sense when spherical support lands. And if we start talking about spherical geometries as geographies, the internal API consistency of geopandas falls apart.

I might miss something but I don't really see any critical or unsolvable API issue here?

What would prevent extending GeoPandas API with explicit `set_geography`, `active_geography_name`, `rename_geography` and `geography` keywords and keeping the "geometry" API working with planar geometries? For example, I'd quite like that `GeoDataFrame.from_arrow(table, geometry="geom")` creates a column with shapely objects while `GeoDataFrame.from_arrow(table, geography="geog")` creates a column with spherely objects. I guess it wouldn't be too hard to ensure that a GeoDataFrame has either a "geometry" or a "geography" active column but not both, so we can still have easy and unambiguous access to the geospatial information? It seems that other geospatial tools also took the route of separate geometry vs. geography API accessors (e.g., PostGIS `geometry_columns` and `geography_columns` views).

If the goal is to reuse that same "geometry" API for handling all objects, then I agree "geometry" API with both "geometry" and "geography" types is not ideal. Although maybe not that a big deal either? "geometry" in the API (either as method name or keyword argument name) seem to refer more to the column of spatial features than their actual type, so it wouldn't seem totally unreasonable to me if the geography type was also handled via the "geometry" API as well, assuming that the ratio of planar vs. spherical use cases is more on the order of 90/10 than 50/50... This minor API inconsistency in GeoPandas might be worth the larger consistency with other geospatial tools?

> We also discussed a proposal to make then change in spherely to stick closer to the S2 terminology rather than introducing the term geography

So s2geography uses the term Geography to represent the "simple-features" abstract layer built on top of s2geometry, and for spherely we would use another term closer to S2 for that same abstract layer? That seems very confusing to me, although maybe the proposal you discussed suggests other changes..

**Comment by jorisvandenbossche** [comment]
> What would prevent extending GeoPandas API with explicit `set_geography`, `active_geography_name`, `rename_geography` and `geography` keywords and keeping the "geometry" API working with planar geometries?

Nothing prevents us doing that of course, except for our desire to not have such duplication of methods/attributes/keyword names, I think. 

> > We also discussed a proposal to make then change in spherely to stick closer to the S2 terminology rather than introducing the term geography
> 
> So s2geography uses the term Geography to represent the "simple-features" abstract layer built on top of s2geometry, and for spherely we would use another term closer to S2 for that same abstract layer? That seems very confusing to me, although maybe the proposal you discussed suggests other changes..

Yes, I indeed did not yet open that issue to bring this up to spherely. I don't think this will be that confusing for users though, since most of them will not known that we use `s2geography` layer on top of `s2geometry`, that's mostly an implementation detail (I agree it looks a bit odd from a contributor point of view, but then you could also argue that `s2geography` is a confusing name since this is a simple features layer on top of s2geometry, but there is nothing in that package that makes it more "geography" than s2geometry itself ;))

**Comment by jorisvandenbossche** [comment]
> Yes, I indeed did not yet open that issue to bring this up to spherely.

Opened https://github.com/benbovy/spherely/issues/100
